hyein/2020-07-30-optional.md

```diff
@@ -0,0 +1,147 @@
+# Optional 기초 : NPE에서 벗어나보자.
+
+## 들어가며
+- WAS log에서 NullPointerException을 발견했던 과거
+  - 발생 시점을 추적하기가 어려웠다. 스택 트레이스 보는데도 모르겠음.
+- 메서드를 짜는데 객체가 없을 경우 null을 리턴 x 100(반복)
+  - null 체크를 계속 해야 하는 불편. 코드가 거슬린다. (if 널 존재여부...) 까먹었을까봐 불안하다.
+- 그러던 중 스프링 강의에서 객체를 Optional로 감싸 NPE를 막는 걸 발견
+    - => 배워야겠다.
+
+ Optional은 람다, 스트림을 알아야 더 사용하기 좋지만...아직 사용하지 못한다. 그저 Optional만 알았을 뿐. 때문에 람다와 스트림을 몰라도 써먹을 수 있는 Optional의 기초만 정리해보도록 한다.
+
+
+## Optional<T>
+ Optional은 java8부터 등장했다. Optional<T>은 지네릭 클래스로 'T타입의 객체'를 감싸는 래퍼 클래스다. T 타입이기 때문에 Optional 타입의 객체는 모든 타입의 참조변수를 담을 수 있다. 
+
+```java
+public final class Optional<T> {
+    private final T value; // T 타입의 참조 변수
+}
+```
+
+### 객체 생성과 초기화
+Opitonal 객체를 생성할 때는 `of()` 또는 `ofNullable()`을 사용한다. **참조변수의 값이 null일 가능성이 있으면 ofNullable을 사용해야 한다.** of를 쓰면 NullPointerException이 발생하기 때문이다.  
+
+1. 생성
+```java
+Optional<String> optVal = Optional.of(null); //NPE 발생
+Optional<String> optVal = Optional.ofNullable(null) //ok
+```
+
+```java
+/* 사용 예: MemberRepository */
+private static Map<Long, Member> store = new HashMap<>();
+
+public Optional<Member> findById(Long id) {
+	return Optional.ofNullable(store.get(id)); //ok
+    // return Optional.of(store.get(id));
+}
+```
+-  id(key) 값으로 가져온 Member(value) 객체가 없을 수도 있으므로 Optional.of보다  Optional.ofNullable을 사용한다.
+
+
+이 부분을 정리하면서 '그렇다면 무조건 `ofNullable()`만 사용하면 되지 않나? `of()`는 왜 쓰지?'라는 궁금증이 들었다. 스터디원 분이 알려주신 블로그 글에서 그 해답을 찾을 수 있었다. 아래는 블로그 글에 나온 Bad/Good 코드 예시다.
+
+```java
+// Bad
+return Optional.of(member.getEmail());  // member의 email이 null이면 NPE 발생
+
+// Good
+return Optional.ofNullable(member.getEmail());
+```
+email같은 경우 null일 가능성이 존재한다. 이렇게 null이 될 가능성이 조금이라도 있으면 `ofNullable()`을 쓴다.
+
+```java
+// Bad
+return Optional.ofNullable("READY");
+
+// Good
+return Optional.of("READY");
+```
+
+`of()`는 null이 발생할 리 없는 변수에서 사용한다. "READY"같은 경우 null이 될 리가 없으니 `ofNullable()`보다 `of()`를 쓴다. 
```

Comment:
null 이 발생할리 없는 변수인데 Optional로 래핑을 해야할까요?

Reply:
메서드 반환 타입이 Optional인 경우 래핑은 선택이 아니라 필수입니다!  위 예시는 String 타입의 변수가 들어왔을 때 이 변수가 null일 가능성이 조금이라도 있다면 -> ofNullable()  / null일 가능성이 없다면 -> of()를 권장하는 코드인데요

`public static <T> Optional<T> ofNullable(T value) {
    return value == null ? empty() : of(value); 
}`  

ofNullable 메서드 내부를 살펴보면 null 체크 코드가 들어있습니다. 굳이 null이 발생할 리 없는 value인데 of가 아닌 ofNullable을 사용할 이유가 없다는거죠.  

프로그램 로직에서 null이 발생할 리가 없는데 발생한다? -> 그건 버그로 간주하고 NPE를 던지는게 설계상 맞다고 합니다. 

> Because, as others have said in answers, Optional is designed to reduce NPEs at the method caller level, not necessarily at the method implementation level. Optional.of is for when one knows the value is not null (and an error should be thrown if it is), Optional.empty is for when one knows the value is null, and Optional.ofNullable is for when one is not sure if the value is null or not

stackoverflow에서 찾은 답변인데요. 내용을 정리하면,

Optional은 메소드 구현 레벨이 아니라 메소드 호출자 레벨에서 NPE를 줄이도록 설계되었음
- of => value가 not null임을 알고 있을 때
- empty => value가 null임을 알고 있을 때
- ofNullable -> value가 null인지 not null인지 불확실할 때 
 
더 다양한 설명이 필요하면 [stackoverflow](https://stackoverflow.com/questions/53810048/java-8-optional-why-of-and-ofnullable?noredirect=1&lq=1) 를 참고하시면 좋을거 같아요. 

아, 그리고 혹시 질문의 요지가 필드 래핑 여부라면 Optional은 주로 메서드 반환 타입으로 사용하고 필드에서의 사용은 권장하지 않는다고 합니다! 

간단한(?) 질문인데 말문이 막혀서,,,크흠흠ㅎ믐.. 이것저것 더 찾아봤네요ㅠㅠㅠ 만족할 만한 답변이셨으면 좋겠습니다.

Reply:
조금더 첨언드리자면 반환 타입이 Optional<T> 일 경우에 해당 API를 이용하는 개발자는 해당 메서드가 null을 반환할 수 있다는 가능성을 인지할 수 있기 때문에 Optional로 래핑합니다!

 반환 타입이 Optional인 메서드의 내부에서는 결과값을 반환할 때 결과로 줄 값이 null일 수도 있고, 아닐 수도 있습니다. 그러면 무조건 ofNullable() 을 이용하면 될 것을 왜 굳이 of()를 쓰는거지? 저도 의문을 가졌었는데요. 이는 코드 레벨에서 해당 값이 null임을 언제 인지할 수 있는지에 대한 차이로 인해 그런 것 같습니다. 만약에 무조건 ofNullable()을 사용한다고 해보겠습니다. 만약 메서드의 반환값이 null이었다면 반환된 Optional에는 null 들어있겠지요. 해당 값은 .get()을 호출해도 익셉션이 발생하지 않을 것입니다. 왜냐하면 ofNullable()을 통해 null이 들어갈 수 있음을 이미 시사했기 때문이지요. 여기서 문제가 발생할 수 있습니다. 개발자가 이를 인지하지 못하고 테스트 코드를 돌렸고, 우연하게 null이 들어가있음에도 특정 로직에 의해 통과됐습니다. null 값을 이용하고 있음에도 말이죠! 그런 상태로 배포가 되었다고 생각해보시면 아마 끔찍한 결과가 발생할 것임을 예상할 수 있겠죠.

 만약 of() 메서드를 사용했다고 생각해보겠습니다. 개발자가 이 메서드를 호출했고, 반환된 Optional에 get() 메서드를 호출하는 순간 NoSuchElementException이 발생할 것입니다. 이로 인해 개발자는 테스트 단계에서 결함을 발견하고 신속하게 수정할 수 있겠지요!

 따라서 특정 상황에 따라서는 반드시 null이 반환이 안되는 상황에서 of()를 통해 에러 상황을 조금 더 빨리 캐치할 수 있겠지요!

 제가 생각했을 때는 위와 같은 이유로 of()를 통해 NPE의 위험을 조금이나마 줄이고자 하는 이유가 있지 않을까 싶습니다.

 제가 생각한 부분이 맞을지는 모르겠는데 다른 이유가 있다고 생각하시는 분들은 코멘트 주시면 감사드리겠습니다 :)